File: src/commands/update.ts
import { Command } from "@effect/cli"
import { Console, Effect } from "effect"
import { UpdaterService } from "../services/updater.js"
import { VERSION } from "../version.js"
import { bold, dim, green } from "../fmt.js"

export const updateCommand = Command.make("update", {}, () =>
  Effect.gen(function* () {
    const updater = yield* UpdaterService
    yield* Console.log(`${dim("Current:")} ${VERSION}`)

    const latest = yield* updater.fetchLatestVersion()
    if (latest === VERSION) {
      yield* Console.log(`${dim("Latest: ")} ${latest}  ${dim("(up to date)")}`)
      return
    }

    yield* Console.log(`${dim("Latest: ")} ${bold(latest)}  ${dim("(updating…)")}`)
    yield* updater.installLatest(latest)
    yield* Console.log(green(`✓ Updated to ${latest}.`))
    yield* Console.log(dim("Run 'ship --version' to confirm."))
  }).pipe(
    Effect.catchAll((e) => Console.error(`Error: ${e.message}`))
  )
)

File: src/commands/update.rs
use crate::errors::Result;
use crate::fmt::{bold, dim, green};
use crate::services::updater;
use crate::version::VERSION;

// ---------------------------------------------------------------------------
// ship update — download and install the latest release
// ---------------------------------------------------------------------------

pub fn run() {
    if let Err(e) = run_inner() {
        eprintln!("Error: {e}");
    }
}

fn run_inner() -> Result<()> {
    println!("{} {}", dim("Current:"), VERSION);

    let latest = updater::fetch_latest_version()?;
    if latest == VERSION {
        println!("{} {}  {}", dim("Latest: "), latest, dim("(up to date)"));
        return Ok(());
    }

    println!(
        "{} {}  {}",
        dim("Latest: "),
        bold(&latest),
        dim("(updating…)")
    );
    updater::install_latest(&latest)?;
    println!("{}", green(format!("✓ Updated to {latest}.")));
    println!("{}", dim("Run 'ship --version' to confirm."));
    Ok(())
}
